pkg/proxy/iptables/endpoints_tracker.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptables
+
+import (
+	"net"
+	"reflect"
+	"strconv"
+	"sync"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/proxy"
+	"k8s.io/kubernetes/pkg/proxy/config"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/slice"
+)
+
+// internal state
+type endpointsTrackerState struct {
+	endpoints []string
+}
+
+// EndpointsTracker is used by ProxierIptables as a receiver of endpoint changes.
+type EndpointsTracker struct {
+	lock     sync.Mutex
+	services map[proxy.ServicePortName]*endpointsTrackerState
+}
+
+var _ config.EndpointsConfigHandler = &EndpointsTracker{}
+
+func (et *EndpointsTracker) GetService(service proxy.ServicePortName) (*endpointsTrackerState, bool) {
+	et.lock.Lock()
+	defer et.lock.Unlock()
+	info, exists := et.services[service]
+	return info, exists
+}
+
+// Returns a new EndpointsTracker
+func NewEndpointsTracker() *EndpointsTracker {
+	return &EndpointsTracker{
+		services: make(map[proxy.ServicePortName]*endpointsTrackerState),
+	}
+}
+
+// expects that et.lock is already held, for internal use
+func (et *EndpointsTracker) newServiceInternal(svcPort proxy.ServicePortName) *endpointsTrackerState {
+	if _, exists := et.services[svcPort]; !exists {
+		et.services[svcPort] = &endpointsTrackerState{}
+		glog.V(4).Infof("EndpointsTracker service %q did not exist, created", svcPort)
+	}
+	return et.services[svcPort]
+}
+
+// adds a new service to the EndpointsTracker
+func (et *EndpointsTracker) NewService(svcPort proxy.ServicePortName) {
+	et.lock.Lock()
+	defer et.lock.Unlock()
+	et.newServiceInternal(svcPort)
+}
+
+// OnUpdate takes in a slice of updated endpoints.
+// because all actual load-balancing is done in ProxierIptables we just keep
+// track of the endpoints and map them to services
+// We can't just handle OnUpdate in ProxierIptables because Proxier already has a
+// clashing OnUpdate function.
+func (et *EndpointsTracker) OnUpdate(allEndpoints []api.Endpoints) {
```

Comment:
It used to track diffs to handle removed endpoints. That's done by finding
unused rules now.
All of this is going to get rewritten more cleanly now that we've decided
to pare this down to just portals to start with no nodeports or other
special cases.
I need to discuss with @thockin but may just spin this out as a smaller
cleaned up PR with less commits.
On Aug 7, 2015 6:25 PM, "Abhi Shah" <notifications@github.com> wrote:

> In pkg/proxy/iptables/endpoints_tracker.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210#discussion_r36566767>
> :
>
> > +	return et.services[svcPort]
> > +}
> > +
> > +// adds a new service to the EndpointsTracker
> > +func (et *EndpointsTracker) NewService(svcPort proxy.ServicePortName) {
> > +	et.lock.Lock()
> > +	defer et.lock.Unlock()
> > +	et.newServiceInternal(svcPort)
> > +}
> > +
> > +// OnUpdate takes in a slice of updated endpoints.
> > +// because all actual load-balancing is done in ProxierIptables we just keep
> > +// track of the endpoints and map them to services
> > +// We can't just handle OnUpdate in ProxierIptables because Proxier already has a
> > +// clashing OnUpdate function.
> > +func (et *EndpointsTracker) OnUpdate(allEndpoints []api.Endpoints) {
>
> Is there a reason why we just don't replace the entire services map
> instead of trying to do diffs etc ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9210/files#r36566767>
> .
>